api-fetch: Add Accept json header to l10n requests

TimothyBJacobs opened the pull request:
## Description
Fixes #11327. The server side localization switching requires an 'Accept' or
'Content-Type' header set to 'application/json` to detect that it is a JSON
request.

## How has this been tested?
Manually verify that `Accept: application/json` is in the `apiFetch` requests and the UI is properly localized.
With a unit test verifying that the `headers` option is mutated as expected.

## Screenshots <!-- if applicable -->
![fix localization](https://user-images.githubusercontent.com/3460448/47821146-15e37200-dd36-11e8-8ad1-87ac673de913.png)

## Types of changes
Bug Fix

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

I'm not sure if this can be properly unit tested. I considered trying to check that the `headers` passed into options is properly mutated, but my local install is currently failing when trying to run tests.

TimothyBJacobs committed:
api-fetch: Add Accept json header to l10n requests

Fixes #11327. The server side localization switching requires an 'Accept' or
'Content-Type' header set to 'application/json` to detect that it is a JSON
request.

TimothyBJacobs committed:
Add tests to check if the accept header is properly applied.

youknowriad commented on packages/api-fetch/src/index.js:
```diff
@@ -36,6 +36,10 @@ function apiFetch( options ) {
 		}
 
+		if ( ! headers.Accept ) {
+			headers.Accept = 'application/json';
```
Are all api-fetch requests `application/json` requests? I'm not sure about that? One I can think of is the metabox save request. I wonder if there's others?

Why is it necessary for the server side localization switching to have an 'Accept' header?

TimothyBJacobs replied:
> Are all api-fetch requests application/json requests? I'm not sure about that? One I can think of is the metabox save request. I wonder if there's others?

I'm not sure. I proposed this, or doing it in the locale middleware.

> Why is it necessary for the server side localization switching to have an 'Accept' header?

The server side localization requires the Accept header because it uses `wp_is_json_request()`. Core does not currently have a method to check whether the current request is a REST request early enough for the locale switching to work and adding a reliable one would be quite difficult.

youknowriad replied:
> The server side localization requires the Accept header because it uses `wp_is_json_request()`.

Maybe a crazy question but can't we just do the switch for all requests :) and not only REST requests? Seems like a `_locale` argument is explicit enough for all requests :) 

I'd understand the hesitation though and would be ok with the change but we should just test it properly. It's also unfortunate because this means we can't have APIs returning something else that is not JSON?

TimothyBJacobs replied:
> Maybe a crazy question but can't we just do the switch for all requests :) and not only REST requests? Seems like a _locale argument is explicit enough for all requests :)

Something for @pento and @danielbachhuber. I image we wouldn't really want to touch front-end localization as that opens up a whole new can of worms, but ¯\\\_(ツ)\_/¯.

>  It's also unfortunate because this means we can't have APIs returning something else that is not JSON?

We could adjust the `Accept` header slightly to `Accept: application/json, */*;q=0.1` that would indicate we'd prefer a JSON response, but fallback to everything. `wp_is_json_request` just checks that `application/json` is present in the accept header.

danielbachhuber replied:
> > The server side localization requires the Accept header because it uses wp_is_json_request().
> 
> Maybe a crazy question but can't we just do the switch for all requests :) and not only REST requests? Seems like a _locale argument is explicit enough for all requests :)

This probably isn't a good idea.

aduth replied:
> Are all api-fetch requests `application/json` requests? I'm not sure about that? One I can think of is the metabox save request. I wonder if there's others?

The REST API is a JSON API, so I personally think it's perfectly reasonable to say "Yes, all requests are `application/json`" (or, more realistically, "should be").

youknowriad replied:
The question is whether `apiFetch` is limited to the REST API (which is not the case in our usage for the moment).

TimothyBJacobs committed:
Widen the Accept header to include */*.

This indicates that JSON is preferred  and is enough for the server to
determine that it is a JSON request.

aduth commented on packages/api-fetch/src/index.js:
```diff
@@ -36,6 +36,10 @@ function apiFetch( options ) {
 			headers[ 'Content-Type' ] = 'application/json';
 		}
 
+		if ( ! headers.Accept ) {
```
Minor: Rather than adding a bunch of `if` statements, we could simplify this using [`Object.assign`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign) in assigning these as defaults:

```js
const headers = {
	Accept: 'application/json, */*;q=0.1',
	'Content-Type': 'application/json',
	...remainingOptions.headers
};
```

TimothyBJacobs replied:
Makes sense, will update.

aduth commented on packages/api-fetch/src/index.js:
```diff
@@ -36,6 +36,10 @@ function apiFetch( options ) {
 		}
 
+		if ( ! headers.Accept ) {
+			headers.Accept = 'application/json, */*;q=0.1';
```
Do we _want_ the "any" fallback?

TimothyBJacobs replied:
Sorry, I thought that was resolved in the message with @youknowriad. Previously, `*/*` would be sent. Since I guess `apiFetch` isn't only intended for use with JSON APIs, I think it makes sense to use `*/*` as a fallback. Particularly since they can disable `parse` to allow them to handle their own responses as necessary, meaning that requiring JSON would no longer be true.

aduth replied:
>Since I guess `apiFetch` isn't only intended for use with JSON APIs

But I shall prove that it is _!_

Since the API Fetch module is a ["utility to make WordPress REST API requests"](https://github.com/WordPress/gutenberg/tree/master/packages/api-fetch) and the REST API operates by ["sending and receiving JSON"](https://developer.wordpress.org/rest-api/) we can conclude that — via transitive relation — the API Fetch module sends and receives JSON.

TimothyBJacobs replied:
I don't have any strongly held opinion here, just going off of @youknowriad's feedback.

TimothyBJacobs commented:
Hmm. It's working for me.
![screen shot 2018-11-01 at 1 41 19 pm](https://user-images.githubusercontent.com/3460448/47868924-eab35e00-dddb-11e8-911a-d34a30cd4619.png)
![screen shot 2018-11-01 at 1 41 32 pm](https://user-images.githubusercontent.com/3460448/47868925-eab35e00-dddb-11e8-9b2c-881817a3df2e.png)
![screen shot 2018-11-01 at 1 41 36 pm](https://user-images.githubusercontent.com/3460448/47868926-eab35e00-dddb-11e8-84c0-63944d31292e.png)

aduth commented:
@danielbachhuber Have you downloaded language packs for the plugin (checking Updates screen)?

danielbachhuber commented:
@TimothyBJacobs Ok, I trust you. Might be a browser cache or something else on my end.

TimothyBJacobs committed:
Use assign to simplify applying headers.

TimothyBJacobs commented:
I double checked and did a fresh build, it seems to be working properly. Given how fragile the whole locale system works, it'd be nice if someone could give it another check to make sure I'm not the odd one out 😄.

I adjusted it to use the spread assign syntax. This breaks the tests, which were probably too brittle anyways.

danielbachhuber commented:
> Given how fragile the whole locale system works, it'd be nice if someone could give it another check to make sure I'm not the odd one out 😄.

I was the odd one out: I had Gutenberg the plugin deactivated 😭

youknowriad commented on packages/api-fetch/src/index.js:
```diff
@@ -31,10 +31,11 @@ function checkCloudflareError( error ) {
-		}
+		const headers = {
+			Accept: 'application/json, */*;q=0.1',
+			'Content-Type': 'application/json',
```
Image uploads are not JSON, this could have broke image uploads in media blocks.

danielbachhuber replied:
@youknowriad Sorry, not sure I follow. What's your point?

Is it:

* That we should have e2e tests covering image uploads?
* That we should have unit tests in this package ensuring multiple content types are supported?
* Something else?

youknowriad replied:
Just that uploading images is broken in master now because of this :)

aduth replied:
I'm working on a fix currently.

aduth replied:
Fix at #11437
